libs/utils/src/celix_convert_utils.c

```diff
@@ -107,25 +112,281 @@ long celix_utils_convertStringToLong(const char* val, long defaultValue, bool* c
+                                                           celix_status_t (*addEntry)(celix_array_list_t*,
+                                                                                      const char*)) {
+    assert(list != NULL);
+    *list = NULL;
+
+    if (!val && defaultValue) {
+        *list = celix_arrayList_copy(defaultValue);
+        return *list ? CELIX_ILLEGAL_ARGUMENT : CELIX_ENOMEM;
+    } else if (!val) {
+        return CELIX_ILLEGAL_ARGUMENT;
+    }
+
+    celix_array_list_create_options_t opts = CELIX_EMPTY_ARRAY_LIST_CREATE_OPTIONS;
+    opts.simpleRemovedCallback = freeCb;
+    celix_autoptr(celix_array_list_t) result = celix_arrayList_createWithOptions(&opts);
+    if (!result) {
+        return CELIX_ENOMEM;
+    }
+
+    char* buf = NULL;
+    size_t bufSize = 0;
+    FILE* entryStream = NULL;
+    celix_status_t status = CELIX_SUCCESS;
+    size_t max = strlen(val);
+    for (size_t i = 0; i <= max; ++i) {
+        if (!entryStream) {
+            entryStream = open_memstream(&buf, &bufSize);
+            if (!entryStream) {
+                 return CELIX_ENOMEM;
+            }
+        }
+        if (val[i] == ESCAPE_CHAR) {
+            // escape character, next char must be escapeChar or separatorChar
+            if (i + 1 < max && (val[i + 1] == ESCAPE_CHAR || val[i + 1] == SEPARATOR_CHAR)) {
+                // write escaped char
+                i += 1;
+                int rc = fputc(val[i], entryStream);
+                if (rc == EOF) {
+                    return CELIX_ENOMEM;
+                }
+                continue;
+            } else {
+                // invalid escape (ending with escapeChar or followed by an invalid char)
+                status = CELIX_ILLEGAL_ARGUMENT;
+                break;
+            }
+        } else if (val[i] == SEPARATOR_CHAR || val[i] == '\0') {
+            //end of entry
+            fclose(entryStream);
+            entryStream = NULL;
+            status = addEntry(result, buf);
+            if (status == CELIX_ENOMEM) {
+                return status;
+            }
+        } else {
+            //normal char
+            int rc = fputc(val[i], entryStream);
+            if (rc == EOF) {
+                return CELIX_ENOMEM;
+            }
+        }
+    }
+
+
+    if (status == CELIX_SUCCESS) {
+        *list = celix_steal_ptr(result);
+    } else if (status == CELIX_ILLEGAL_ARGUMENT) {
+        if (defaultValue) {
+            *list = celix_arrayList_copy(defaultValue);
+            return *list ? status : CELIX_ENOMEM;
         }
+        return status;
+    }
+    return status;
+}
+
+celix_status_t celix_utils_addLongEntry(celix_array_list_t* list, const char* entry) {
+    bool converted;
+    long l = celix_utils_convertStringToLong(entry, 0L, &converted);
+    if (!converted) {
+        return CELIX_ILLEGAL_ARGUMENT;
+    }
+    return celix_arrayList_addLong(list, l);
+}
+
+celix_status_t celix_utils_convertStringToLongArrayList(const char* val,
```

Comment:
As external API, `celix_utils_convertStringToXXXArrayList` all suffer from lack of type checking for `defaultValue`.

Reply:
I think the main difficulty of replacing `celix_utils_convertStringToXXXArrayList` by `celix_utils_convertStringToArrayList` is that the string contains no typing information, unlike JSON.